test/Test.ts

```diff
@@ -14,222 +13,336 @@ let sarcoDao: SignerWithAddress;
+          balanceVoter1Before.add(voterReward)
+        );
+        expect(await sarco.balanceOf(voter2.address)).to.equal(
+          balanceVoter2Before.add(voterReward)
+        );
+      });
+
+      it('should reset the voters internal accounting mapping balanceOf[]', async () => {
+        await collection.connect(voter1).claim();
+        await collection.connect(voter2).claim();
+
+        // check internal accounting clears balances of voters once rewards are claimed
+        expect(await collection.balanceOf(voter1.address)).to.equal(zero);
+        expect(await collection.balanceOf(voter1.address)).to.equal(zero);
+      });
+
+      it('should reset total claimable tokens by voters once all have been claimed', async () => {
+        await collection.connect(voter1).claim();
+        await collection.connect(voter2).claim();
+
+        // check internal accounting clears total claimable tokens by voters
+        expect(await collection.claimableByVoters()).to.equal(zero);
+      });
+
+      it('emit Claim', async () => {
+        expect(await collection.connect(voter1).claim()).to.emit(voterReward, 'Claim');
+      });
+    });
 
-        beforeEach(async () => {
-          let rewards: Reward[] = [{_address: voter1.address, _amount: voterReward}, {_address: voter2.address, _amount: voterReward}]
-          await collection.connect(sarcoDao).allocateRewards(rewards)
-        });
-
-        it("should revert when function not called by owner", async () => {
-          await expect(collection.connect(voter1).daoWithdraw())
-          .to.be.revertedWith('Ownable: caller is not the owner')
-        })
-
-        it("should revert if all rewards have been allocated", async () => {
-          const toBeAllocated = await collection.unallocatedRewards()
-          // setting voter rewards as what can still be allocated to voters, divided by number of voters
-          voterReward = toBeAllocated.div(2);
-          let rewards: Reward[] = [{_address: voter1.address, _amount: voterReward}, {_address: voter2.address, _amount: voterReward}]
+    context('Successfully claim the SARCO randomly allocated to 50 voters', () => {
+      it('SARCO balance of each voter should increase by their allocated reward and internal accounting is adjusted', async () => {
+        let votersNumber = 50;
+        // helper function to distribute rewards randomly among a number of voters
+        let rewards = await randomRewards(votersNumber, initialContractBalance);
+        await collection.connect(sarcoDao).allocateRewards(rewards);
+
+        // voters claim their rewards if they have been allocated any
+        for (let i = 0; i < rewards.length; i++) {
+          const balanceVoterBefore = await sarco.balanceOf(rewards[i]._address);
+          let signer = rewards[i]._signer as SignerWithAddress;
+
+          if (rewards[i]._amount.eq(zero)) {
+            await expect(collection.connect(signer).claim()).to.be.revertedWith(
```

Comment:
Probably a good idea to manually push a reward with zero value to `rewards` before allocating, so that this conditional is guaranteed to run at least once.

Reply:
We decided to invert the if statement and remove the expect revert condition